sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SymbolicExecution/BaseExplodedGraph.cs

```diff
@@ -1,442 +1,442 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+
+namespace SonarAnalyzer.Helpers.FlowAnalysis.Common
+{
+    internal abstract class BaseExplodedGraph
+    {
+        internal const int MaxStepCount = 1000;
+        internal const int MaxInternalStateCount = 10000;
+        private const int MaxProgramPointExecutionCount = 2;
+
+        private readonly List<ExplodedGraphNode> nodes = new List<ExplodedGraphNode>();
+        private readonly Dictionary<ProgramPoint, ProgramPoint> programPoints = new Dictionary<ProgramPoint, ProgramPoint>();
+        private readonly Queue<ExplodedGraphNode> workList = new Queue<ExplodedGraphNode>();
+        private readonly HashSet<ExplodedGraphNode> nodesAlreadyInGraph = new HashSet<ExplodedGraphNode>();
+
+        private readonly IControlFlowGraph cfg;
+        private readonly ISymbol declaration;
+        private readonly IEnumerable<IParameterSymbol> declarationParameters;
+        private readonly IEnumerable<IParameterSymbol> nonInDeclarationParameters;
+        private readonly Common.LiveVariableAnalysis lva;
+
+        protected readonly ICollection<ExplodedGraphCheck> explodedGraphChecks = new List<ExplodedGraphCheck>();
+
+        internal SemanticModel SemanticModel { get; }
+
+        public event EventHandler ExplorationEnded;
+        public event EventHandler MaxStepCountReached;
+        public event EventHandler MaxInternalStateCountReached;
+        public event EventHandler<InstructionProcessedEventArgs> InstructionProcessed;
+        public event EventHandler<VisitCountExceedLimitEventArgs> ProgramPointVisitCountExceedLimit;
+        public event EventHandler ExitBlockReached;
+        public event EventHandler<ConditionEvaluatedEventArgs> ConditionEvaluated;
+
+        protected BaseExplodedGraph(IControlFlowGraph cfg, ISymbol declaration, SemanticModel semanticModel, LiveVariableAnalysis lva)
+        {
+            this.cfg = cfg;
+            this.declaration = declaration;
+            this.lva = lva;
+
+            SemanticModel = semanticModel;
+
+            declarationParameters = declaration.GetParameters();
+            nonInDeclarationParameters = declarationParameters.Where(p => p.RefKind != RefKind.None);
+        }
+
+        public void Walk()
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 17 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS3776)